File: packages/ui-compose/ui/src/test/kotlin/com/kinetixui/ui/blocks/TeamMembersBlockTest.kt
package com.kinetixui.ui.blocks

import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.test.junit4.v2.createComposeRule
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.unit.dp
import com.kinetixui.ui.KinetixAvatar
import com.kinetixui.ui.KinetixAvatarFallback
import com.kinetixui.ui.KinetixButton
import com.kinetixui.ui.KinetixButtonSize
import com.kinetixui.ui.KinetixButtonVariant
import com.kinetixui.ui.KinetixCard
import com.kinetixui.ui.KinetixCardHeader
import com.kinetixui.ui.KinetixCardTitle
import com.kinetixui.ui.KinetixList
import com.kinetixui.ui.KinetixListItem
import com.kinetixui.ui.KinetixSeparator
import com.kinetixui.ui.KinetixTheme
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

// The "Team members" block — see TestimonialBlockTest.kt for how block fixtures work.
//
// kx-block:start
data class TeamMember(val name: String, val role: String, val initials: String)

@Composable
fun TeamMembersBlock(
    team: List<TeamMember> = listOf(
        TeamMember("Ada Lovelace", "Owner", "AL"),
        TeamMember("Grace Hopper", "Admin", "GH"),
        TeamMember("Alan Turing", "Member", "AT"),
    ),
    onRemove: (TeamMember) -> Unit = {},
) {
    KinetixCard(Modifier.width(360.dp)) {
        KinetixCardHeader { KinetixCardTitle("Team") }
        KinetixSeparator()
        KinetixList {
            team.forEach { member ->
                KinetixListItem(
                    title = member.name,
                    description = member.role,
                    leading = { KinetixAvatar { KinetixAvatarFallback(member.initials) } },
                    trailing = {
                        KinetixButton(
                            { onRemove(member) },
                            variant = KinetixButtonVariant.Ghost,
                            size = KinetixButtonSize.Sm,
                        ) { Text("Remove") }
                    },
                )
            }
        }
    }
}
// kx-block:end

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [34])
class TeamMembersBlockTest {
    @get:Rule
    val rule = createComposeRule()

    @Test
    fun renders_every_member_with_their_role() {
        rule.setContent { KinetixTheme(darkTheme = false) { TeamMembersBlock() } }
        rule.onNodeWithText("Team").assertExists()
        rule.onNodeWithText("Ada Lovelace").assertExists()
        rule.onNodeWithText("Owner").assertExists()
        rule.onNodeWithText("Alan Turing").assertExists()
    }
}

File: packages/ui-flutter/test/blocks/team_members_block_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:kinetix_ui/kinetix_ui.dart';

// The "Team members" block — see testimonial_block_test.dart for how block fixtures work.
//
// kx-block:start
class TeamMember {
  const TeamMember(this.name, this.role, this.initials);

  final String name;
  final String role;
  final String initials;
}

class TeamMembersBlock extends StatelessWidget {
  const TeamMembersBlock({super.key, this.team = defaultTeam, this.onRemove});

  static const List<TeamMember> defaultTeam = <TeamMember>[
    TeamMember('Ada Lovelace', 'Owner', 'AL'),
    TeamMember('Grace Hopper', 'Admin', 'GH'),
    TeamMember('Alan Turing', 'Member', 'AT'),
  ];

  final List<TeamMember> team;
  final void Function(TeamMember)? onRemove;

  @override
  Widget build(BuildContext context) {
    return KinetixCard(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: <Widget>[
          const KinetixCardHeader(children: <Widget>[KinetixCardTitle('Team')]),
          const KinetixSeparator(),
          KinetixList(
            children: <Widget>[
              for (final TeamMember member in team)
                KinetixListItem(
                  title: member.name,
                  description: member.role,
                  leading: KinetixAvatar(child: KinetixAvatarFallback(member.initials)),
                  trailing: KinetixButton(
                    onPressed: onRemove == null ? null : () => onRemove!(member),
                    variant: KinetixButtonVariant.ghost,
                    size: KinetixButtonSize.sm,
                    child: const Text('Remove'),
                  ),
                ),
            ],
          ),
        ],
      ),
    );
  }
}
// kx-block:end

void main() {
  testWidgets('renders every member with their role', (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: KinetixTheme(
          brightness: Brightness.light,
          child: const Scaffold(body: SingleChildScrollView(child: TeamMembersBlock())),
        ),
      ),
    );

    expect(find.text('Team'), findsOneWidget);
    expect(find.text('Ada Lovelace'), findsOneWidget);
    expect(find.text('Owner'), findsOneWidget);
    expect(find.text('Alan Turing'), findsOneWidget);
    expect(find.text('Remove'), findsNWidgets(3));
  });
}
